tests/trantest.h

```diff
@@ -745,16 +1126,62 @@ trantest_mqtt_broker_send_recv(trantest *tt)
 
 		// client recv CONNACK msg.
 		So(nng_recvmsg(tt->reqsock, &msg, 0) == 0);
-		So(msg != NULL);
 		So(nng_mqtt_msg_get_packet_type(msg) == NNG_MQTT_CONNACK);
-		
-		conn_param_free(nng_msg_get_conn_param(msg));
+		rcp = nng_msg_get_conn_param(msg);
+
+		// client send sub & server send suback.
+		trantest_mqtt_sub_send(tt->reqsock, client, true);
+		nng_msleep(100);
+		nng_ctx_recv(work->ctx, work->aio);
+		So((rmsg = nng_aio_get_msg(work->aio)) != NULL);
+		So(nng_msg_get_type(rmsg) == CMD_SUBSCRIBE);
+		// decode sub msg and encode suback msg.
+		So((work->sub_pkt = nng_alloc(sizeof(packet_subscribe))) != NULL);
+		memset(work->sub_pkt, '\0', sizeof(packet_subscribe));
+		work->msg = rmsg;
+		work->pid       = nng_msg_get_pipe(work->msg);
+		work->cparam    = nng_msg_get_conn_param(work->msg);
+		work->proto_ver = conn_param_get_protover(work->cparam);
+		So(decode_sub_msg(work) == 0);
+		So(encode_suback_msg(rmsg, work) == 0);
+		nng_msg_set_cmd_type(rmsg, CMD_SUBACK);
+		nng_aio_set_msg(work->aio, rmsg);
+		nng_ctx_send(work->ctx, work->aio);
+
+		// client send pub msg & server send pub msg.
+		trantest_mqtt_pub(tt->reqsock, false);
+		nng_msleep(100);
+		nng_ctx_recv(work->ctx, work->aio);
+		So((rmsg = nng_aio_get_msg(work->aio)) != NULL);
+		So(nng_msg_get_type(rmsg) == CMD_PUBLISH);
+		nng_aio_set_msg(work->aio, rmsg);
+		nng_ctx_send(work->ctx, work->aio);
+
+		//client recv pub msg.
+		trantest_mqtt_sub_recv(tt->reqsock);
+
+		// client send unsub msg and server recv unsub msg.
+		trantest_mqtt_unsub_send(tt->reqsock, client, true);
+		nng_msleep(100);
+		nng_ctx_recv(work->ctx, work->aio);
+		So((rmsg = nng_aio_get_msg(work->aio)) != NULL);
+		So(nng_msg_get_type(rmsg) == CMD_UNSUBSCRIBE);
+		So((work->unsub_pkt = nng_alloc(sizeof(packet_unsubscribe))) != NULL);
+		work->msg = rmsg;
+		So(decode_unsub_msg(work) == 0);
+		So(encode_unsuback_msg(rmsg, work) == 0);
+		nng_aio_set_msg(work->aio, rmsg);
+		nng_ctx_send(work->ctx, work->aio);
+		nng_aio_finish(work->aio, 0);
```

Comment:
Why? It should be finished by nng?

Reply:
> whoops aio appending error

I think the error of aio appending was caused in this line

Reply:
perhaps you are right
but we can fix in another PR